agents/pysmurf_controller/pysmurf_controller.py

```diff
@@ -93,7 +96,26 @@ def __init__(self, agent, args):
         self.prot = None
         self.protocol_lock = TimeoutLock()
 
-    def _run_script(self, script, args, log):
+        self.current_session = None
+
+        if args.monitor_id is not None:
+            self.agent.subscribe_on_start(
+                self._on_session_data,
+                'observatory.{}.feeds.pysmurf_session_data'.format(args.monitor_id),
+            )
+
+    def _on_session_data(self, _data):
+        data, feed = _data
+
+        if self.current_session is not None:
+            if data['id'] == os.environ.get("SMURFPUB_ID"):
+                if isinstance(data, dict):
+                    self.current_session.data.update(data['payload'])
```

Comment:
There's a timing / race condition issue with current_session.  Currently it's not cleared, which means that any messages coming in will be assigned to the most recent session.  But a rule for session.data is that it must not change once the session is closed (when state -> done).  That could be fixed by clearing current_session before returning from the Task/Process.  But then you might lose messages, if they are issued just before the Task/Process ends.  So I think you also need a short buffer time, after the script has ended but before you close the session, to allow any such messages to propagate into session before you close the session.

Reply:
`current_session` is currently closed out at the end of the _run_script function (line 171). I agree with adding a small buffer before closing it though.